working/0546-patterns/patterns.md

```diff
@@ -0,0 +1,924 @@
+    scopes for a function `method()`. If it finds one, it says the code is a
+    call to that. Otherwise, it treats it as `this.method()`.
+
+4.  **Use the surrounding static type context to decide.** Part of type
+    inference is "downwards" type inference where we use known static type of
+    some enclosing code to decide what a piece of syntax means. For example:
+
+    ```dart
+    Set<int> ints = {};
+    Map<String, int> counts = {};
+    ```
+
+    We know the first `{}` is a set and the second a map because of the static
+    type of the variable declaration that those are initializers for.
+
+How do we choose an approach for these pattern matching problems? The options
+are roughly in order. When possible, the safest solution is to come up with
+distinct syntax. That can be difficult if users already have a strong
+expectation for a certain syntax based on existing language features or other
+languages. Novel syntax carries a high unfamiliarity cost. For features like
+pattern matching that are primarily syntax sugar, an ugly distinct syntax can
+nullify the value of the feature.
+
+Barring that, the less context you use the better. There is a sliding scale for
+how much surrounding code the user has to have loaded in their head before they
+can understand the syntax. Grammatical context is lowest&mdash;they literally
+just need to look at the enclosing text. Lexical scoping is next. They have
+to look at surrounding block scopes and know what's imported. Static types are
+the last&mdash;they have to essentially run the type inference engine in their
+head.
+
+In this proposal, we take approach 2: we use the grammatical context. I think
+this makes the right trade-off between familiar syntax and relatively local
+reasoning, at the expense of a slightly more complex somewhat redundant set of
+patterns. This approach was [directly inspired by Swift][swift], so there's
+evidence this is a good balance.
+
+[swift]: https://docs.swift.org/swift-book/ReferenceManual/Patterns.html
+
+## Two kinds of patterns
+
+Patterns generally do one of three things:
+
+1.  Perform some kind of test on an object to decide whether to execute some
+    piece of code or not. In other words, if the object **matches** the pattern,
+    then evaluate the "body" of some surrounding construct. If the object does
+    not match the value, the we say the pattern was **refuted**.
+
+2.  **Bind** a new variable to the value.
+
+3.  Extract data from some object by **destructuring** it and applying
+    subpatterns to the extracted data.
+
+Functional languages lump those three together into a single grammar, "pattern",
+and allow them to be freely mixed. Instead, for Dart, we have two separate
+categories: **matchers** and **binders**. Matchers are all of the refutable
+patterns that can fail to match a value. Binders are the other patterns that can
+never fail and exist to bind new variables.
```

Comment:
> Are they _syntactic_ categories?

Yes.

> They don't have to bind, though: `int _ = 42;` would be valid and not bind anything. It would also be stupid, but maybe less so as a function parameter.

Right. Maybe "binder" isn't the best name. I was trying to avoid "refutable" and "irrefutable" because those are a mouthful, but maybe they're better.


> So, a pattern is effectively a partial function from values to bindings.
> 
> A pattern occurring as a binder must be total on the value that can flow into it.
> A pattern occurring as a matcher can be partial on the values that can flow into it.

Exactly right.